src/Spago/Command/Fetch.purs

```diff
@@ -442,29 +442,37 @@ toAllDependencies = foldl (Map.unionWith (\l _ -> l)) Map.empty
 
 getGitPackageInLocalCache :: forall a. PackageName -> GitPackage -> Spago (Git.GitEnv a) Unit
 getGitPackageInLocalCache name package = do
+  ensureRepoCloned
+
   let localPackageLocation = Config.getPackageLocation name (GitPackage package)
-  tempDir <- mkTemp' (Just $ printJson Config.gitPackageCodec package)
-  logDebug $ "Cloning repo in " <> tempDir
-  Git.fetchRepo package tempDir >>= case _ of
-    Left err -> die err
-    Right _ -> do
-      logDebug $ "Repo cloned. Moving to " <> localPackageLocation
-      FS.mkdirp $ Path.concat [ Paths.localCachePackagesPath, PackageName.print name ]
-      FS.moveSync { src: tempDir, dst: localPackageLocation }
-
-      -- Note: the package might have been cloned with a tag, but we stick the commit hash in the lockfiles
-      -- so we need to make a copy to a location that has the commit hash too.
-      -- So we run getRef here and then do a copy if the ref is different than the original one
-      -- (since it might be a commit to start with)
-      logDebug $ "Checking if we need to copy the package to a commit hash location..."
-      Git.getRef (Just localPackageLocation) >>= case _ of
-        Left err -> die err
-        Right ref -> do
-          when (ref /= package.ref) do
-            let commitHashLocation = Config.getPackageLocation name (GitPackage $ package { ref = ref })
-            logDebug $ "Copying the repo also to " <> commitHashLocation
-            FS.mkdirp $ Path.concat [ Paths.localCachePackagesPath, PackageName.print name ]
-            FS.copyTree { src: localPackageLocation, dst: commitHashLocation }
+  logDebug $ "Copying repo to " <> localPackageLocation
+  FS.mkdirp $ Path.concat [ Paths.localCachePackagesPath, PackageName.print name ]
+  FS.copyTree { src: repoCacheLocation, dst: localPackageLocation }
+
+  logDebug $ "Checking out " <> package.ref
+  Git.checkout { repo: localPackageLocation, ref: package.ref } >>= rightOrDie_
+
+  -- Note: the package might have been cloned with a tag, but we stick the commit hash in the lockfiles
+  -- so we need to make a copy to a location that has the commit hash too.
+  -- So we run getRef here and then do a copy if the ref is different than the original one
+  -- (since it might be a commit to start with)
+  logDebug $ "Checking if we need to copy the package to a commit hash location..."
+  commitHash <- Git.getRef (Just localPackageLocation) >>= rightOrDie
+  when (commitHash /= package.ref) do
+    let commitHashLocation = Config.getPackageLocation name (GitPackage $ package { ref = commitHash })
+    logDebug $ "Copying the repo also to " <> commitHashLocation
+    FS.copyTree { src: localPackageLocation, dst: commitHashLocation }
+  where
+  repoCacheLocation = Path.concat [ Paths.localCacheGitPath, Config.fileSystemCharEscape package.git ]
+
+  ensureRepoCloned = unlessM (FS.exists repoCacheLocation) do
```

Comment:
My understanding of the patch is that things would break down in a scenario where:
- we clone a git repo when it has refs `aaaaaa` and `bbbbbb`, put it in cache then
- the author of the repo then pushes the ref `cccccc` to the repo
- if that ref is referenced in the package set we'd fail the checkout, since we are not fetching the new refs for the repo once it's in cache

Does this make sense?

Reply:
Yes, you're right. I was planning to pull, but totally forgot.

I am going to add a patch for this, but I am at a loss for how to test this behavior. Maybe create a local Git repo on the file system? Would this be good enough?

Reply:
Yeah that's how I would test it as well